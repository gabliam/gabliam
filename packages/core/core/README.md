[![Build Status][build-image]][build-url]
[![NPM version][npm-image]][npm-url]


# Gabliam core

Gabliam is on top of inversify.
Gabliam help you for register all class in inversify.
Gabilliam is modular with plugins.


# Gabliam phase

## Build phase

1. Load all config file
2. Load all file in scanPath of main
	- All decorators on class add metadata and can register this class
	- if an class has @scanPath, add this folder in queue for load
	- load all plugin
3. Bind classes in DI (inversify)
	- call all plugin.bind
4. load all config classes (order: CoreConfig => PluginConfig => config)
	- call all plugin.config for all instance of config class
5. Build
	- call all plugin.build

## Start phase
/!\ Build phase must be passed

1. call all plugin.start

## Stop phase
/!\ Build and start phases must be passed

1. call all plugin.start

## Detroy phase
/!\ Stop phase must be passed

1. call all plugin.detroy

# Samples
Go to [Samples]



[Samples]: <https://github.com/gabliam/gabliam/tree/master/examples>

# License

  MIT

[build-image]: https://img.shields.io/travis/gabliam/gabliam/master.svg?style=flat-square
[build-url]: https://travis-ci.org/gabliam/gabliam
[npm-image]: https://img.shields.io/npm/v/@gabliam/core.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@gabliam/gabliam/core
