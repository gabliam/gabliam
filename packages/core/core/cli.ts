#!/usr/bin/env node
import 'reflect-metadata';
import yargs from 'yargs';
import { StartCommand } from './src/commands/start-command';

yargs
  .usage('Usage $0 <command> [options]')
  .command(new StartCommand())
  .recommendCommands()
  .demandCommand(1)
  .strict()
  .alias('v', 'version')
  .help('h')
  .alias('h', 'help').argv;

require('yargonaut')
  .style('blue')
  .style('yellow', 'required')
  .helpStyle('green')
  .errorsStyle('red');
