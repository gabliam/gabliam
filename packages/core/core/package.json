{
  "name": "@gabliam/core",
  "version": "0.0.0-PLACEHOLDER",
  "description": "Gabliam is on top of inversify",
  "main": "index.js",
  "typings": "index.d.ts",
  "module": "index.js",
  "scripts": {
    "build": "rimraf lib && tsc -p tsconfig.build.json"
  },
  "files": [
    "lib",
    "src",
    "index.js",
    "index.d.ts"
  ],
  "dependencies": {
    "@gabliam/expression": "0.0.0-PLACEHOLDER",
    "@types/joi": "^14.0.1",
    "caller": "^1.0.1",
    "co": "^4.6.0",
    "debug": "^4.1.1",
    "glob": "^7.1.3",
    "graph-data-structure": "^1.8.0",
    "inversify": "^5.0.1",
    "inversify-logger-middleware": "^3.0.0",
    "joi": "^14.3.1",
    "js-yaml": "^3.12.1",
    "lodash": "^4.17.11",
    "shortstop": "^1.0.3",
    "shortstop-handlers": "^1.0.1",
    "tslib": "^1.9.3"
  },
  "devDependencies": {
    "@types/debug": "^4.1.3",
    "@types/fs-extra": "^5.0.4",
    "@types/glob": "^7.1.1",
    "@types/js-yaml": "^3.11.4",
    "@types/lodash": "^4.14.123",
    "fs-extra": "^7.0.1",
    "properties": "^1.2.1",
    "rxjs": "^6.3.3",
    "toml": "^2.3.3"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/gabliam/gabliam.git"
  },
  "author": "David Touzet <dtouzet@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/gabliam/gabliam/issues"
  },
  "homepage": "https://github.com/gabliam/gabliam/tree/master/packages/core#readme",
  "engines": {
    "node": ">=8.9.0"
  },
  "gitHead": "ec328fa0b62f474d470f74a2fc9637838f54091e"
}
