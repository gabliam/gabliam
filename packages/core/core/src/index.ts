import 'reflect-metadata';

export {
  decorate,
  inject,
  injectable,
  interfaces as inversifyInterfaces,
  multiInject,
  named,
  optional,
  postConstruct,
  tagged,
  targetName,
  unmanaged,
} from 'inversify';
export * from './common';
export {
  APP_CONFIG,
  CORE_CONFIG,
  INJECT_CONTAINER_KEY,
  VALUE_EXTRACTOR,
} from './constants';
export * from './container';
export * from './decorator';
export * from './errors';
export * from './gabliam';
export * from './gabliam-options';
export * from './gabliam-utils';
export * from './interfaces';
export * from './joi';
export { configResolver, FileLoader, Loader, Resolver } from './loaders';
export * from './metadatas';
export * from './reflection';
export * from './registry';
export * from './value-extractor';
