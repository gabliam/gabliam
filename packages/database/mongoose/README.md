[![Build Status][build-image]][build-url]
[![NPM version][npm-image]][npm-url]

# mongoose
mongoose plugin for gabliam


# License

  MIT

[build-image]: https://img.shields.io/travis/gabliam/gabliam/master.svg?style=flat-square
[build-url]: https://travis-ci.org/gabliam/gabliam
[npm-image]: https://img.shields.io/npm/v/@gabliam/mongoose.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@gabliam/mongoose
