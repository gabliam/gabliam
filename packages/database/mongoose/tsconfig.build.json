{
  "extends": "../../../tsconfig.base",
  "compilerOptions": {
    "sourceMap": false,
    "outDir": "../../../dist/packages/database/mongoose",
    "baseUrl": ".",
    "rootDir": ".",
    "paths": {
      "@gabliam/*": [
        "../../../dist/packages/broker/*",
        "../../../dist/packages/cache/*",
        "../../../dist/packages/core/*",
        "../../../dist/packages/database/*",
        "../../../dist/packages/web/*",
        "../../../dist/packages/tools/*"
      ]
    }
  },
  "include":[
    "index.ts",
    "src/**/*.ts"
  ]
}
