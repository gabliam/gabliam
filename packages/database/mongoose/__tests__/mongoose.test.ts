import { MongoosePluginTest } from './mongoose-plugin-test';
import {
  Document,
  MongooseConnection,
  MUnit,
  MongooseConnectionManager,
} from '../src';
import { Gabliam } from '@gabliam/core';
import mongoose from 'mongoose';

interface HeroModel {
  name: string;
}

interface PhotoModel {
  name: string;

  description: string;

  fileName: string;
  views: number;
  isPublished: boolean;
}

let appTest: MongoosePluginTest;
beforeEach(async () => {
  appTest = new MongoosePluginTest();
});

afterEach(async () => {
  try {
    await appTest.destroy();
  } catch (e) {}
});

describe('Errors', () => {
  test('without config folder', async () => {
    await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
  });

  test('with bad config', async () => {
    appTest.addConf('application.mongoose', 'bad');
    await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
  });

  test('bad repo', async () => {
    appTest.addConf(
      'application.mongoose.uri',
      'mongodb://127.0.0.1/mongoosetest'
    );
    await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);
    const connection = appTest.gab.container.get(MongooseConnection);
    expect(() => connection.getRepository('HeroBad')).toThrowError();
  });
});

test('with config uri', async () => {
  appTest.addConf(
    'application.mongoose.uri',
    'mongodb://127.0.0.1/mongoosetest'
  );

  @Document('Hero')
  class Hero {
    static getSchema() {
      return new mongoose.Schema({
        name: {
          type: String,
          required: true,
        },
      });
    }
  }

  appTest.addClass(Hero);

  await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);
  const connection = appTest.gab.container.get(MongooseConnection);
  const repo = connection.getRepository<HeroModel>('Hero');
  let res = await repo.findAll();
  expect(res).toMatchSnapshot();
  await repo.create({
    name: 'test',
  });
  res = await repo.findAll();
  expect(res[0].name).toMatchSnapshot();
  await connection.conn.dropDatabase();
});

test('with one connection with name and entity without munit', async () => {
  appTest.addConf('application.mongoose', {
    uri: 'mongodb://127.0.0.1/mongoosetest',
    name: 'test',
  });

  @Document('Hero')
  class Hero {
    static getSchema() {
      return new mongoose.Schema({
        name: {
          type: String,
          required: true,
        },
      });
    }
  }

  appTest.addClass(Hero);

  await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);
  const connection = appTest.gab.container.get(MongooseConnection);
  const repo = connection.getRepository<HeroModel>('Hero');
  let res = await repo.findAll();
  expect(res).toMatchSnapshot();
  await repo.create({
    name: 'testoneconnection',
  });
  res = await repo.findAll();
  expect(res[0].name).toMatchSnapshot();
  await connection.conn.dropDatabase();
});

test('with config 2 database', async () => {
  appTest.addConf('application.mongoose', [
    {
      name: 'connection1',
      uri: 'mongodb://127.0.0.1/mongoosetestconnection1',
    },
    {
      name: 'connection2',
      uri: 'mongodb://127.0.0.1/mongoosetestconnection2',
    },
  ]);

  @MUnit('connection1')
  @Document('Photo')
  class Photo {
    static getSchema() {
      return new mongoose.Schema({
        name: {
          type: String,
          required: true,
        },
        description: {
          type: String,
          required: true,
        },
        fileName: {
          type: String,
          required: true,
        },
        views: Number,
        isPublished: Boolean,
      });
    }
  }
  appTest.addClass(Photo);

  @MUnit('connection2')
  @Document('Hero')
  class Hero {
    static getSchema() {
      return new mongoose.Schema({
        name: {
          type: String,
          required: true,
        },
      });
    }
  }
  appTest.addClass(Photo);
  appTest.addClass(Hero);

  await expect(appTest.gab.buildAndStart()).resolves.toBeInstanceOf(Gabliam);

  const connection = appTest.gab.container.get(MongooseConnectionManager);

  const repo = connection
    .getConnection('connection1')
    .getRepository<PhotoModel>('Photo');
  let res = await repo.findAll();
  expect(res).toMatchSnapshot();
  await repo.create({
    name: 'test',
    description: 'test desc',
    fileName: 'testfile',
    views: 0,
    isPublished: false,
  });
  res = await repo.findAll();
  expect(res[0].name).toMatchSnapshot();

  const repo2 = connection
    .getConnection('connection2')
    .getRepository<HeroModel>('Hero');
  let res2 = await repo2.findAll();
  expect(res2).toMatchSnapshot();
  await repo2.create({
    name: 'test hero',
  });
  res2 = await repo2.findAll();
  expect(res2[0].name).toMatchSnapshot();
  await connection.getConnection('connection1').conn.dropDatabase();
  await connection.getConnection('connection2').conn.dropDatabase();
});

test('must fail when MUnit not found', async () => {
  appTest.addConf('application.mongoose', {
    uri: 'mongodb://127.0.0.1/mongoosetestbadMunit',
  });

  @MUnit('bad')
  @Document('Photo')
  class Photo {
    static getSchema() {
      return new mongoose.Schema({
        name: {
          type: String,
          required: true,
        },
        description: {
          type: String,
          required: true,
        },
        fileName: {
          type: String,
          required: true,
        },
        views: Number,
        isPublished: Boolean,
      });
    }
  }
  appTest.addClass(Photo);
  await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
});

test('must fail when getConnection not found', async () => {
  appTest.addConf('application.mongoose', {
    uri: 'mongodb://127.0.0.1/mongoosetestbadMunit',
  });

  @Document('Photo')
  class Photo {
    static getSchema() {
      return new mongoose.Schema({
        name: {
          type: String,
          required: true,
        },
        description: {
          type: String,
          required: true,
        },
        fileName: {
          type: String,
          required: true,
        },
        views: Number,
        isPublished: Boolean,
      });
    }
  }

  appTest.addClass(Photo);
  await appTest.gab.buildAndStart();
  const connection = appTest.gab.container.get(MongooseConnectionManager);
  await expect(() =>
    connection.getConnection('notfound')
  ).toThrowErrorMatchingSnapshot();
});
