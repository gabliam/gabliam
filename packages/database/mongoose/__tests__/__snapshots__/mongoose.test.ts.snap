// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Errors with bad config 1`] = `[Error: Error for 'application.mongoose' value{"_original":"bad","details":[{"message":"/"value/" must be one of [object, array]","path":[],"type":"alternatives.types","context":{"types":["object","array"],"label":"value","value":"bad"}}]}]`;

exports[`Errors without config folder 1`] = `[MongoConfigIsMandatoryError: mongooseConfiguration is mandatory.]`;

exports[`must fail when MUnit not found 1`] = `[MongoMUnitNotFoundError: MUnit bad not found]`;

exports[`must fail when getConnection not found 1`] = `"Connection notfound not found"`;

exports[`with config 2 database 1`] = `Array []`;

exports[`with config 2 database 2`] = `"test"`;

exports[`with config 2 database 3`] = `Array []`;

exports[`with config 2 database 4`] = `"test hero"`;

exports[`with config uri 1`] = `Array []`;

exports[`with config uri 2`] = `"test"`;

exports[`with one connection with name and entity without munit 1`] = `Array []`;

exports[`with one connection with name and entity without munit 2`] = `"testoneconnection"`;
